### ISSUE TITLE
questions on parameter x_j of message funcation in source code

### ISSUE THREAD

**Issue by YichengDWu** [pyg-team/pytorch_geometric#341]
## ❓ Questions & Help

I still don't understand what the message function parameter x_j means in the source code. In GCNConv, I only see x_j, while in GATConv there are x_i and x_j. I studied it for hours without understanding it

**Comment by rusty1s** [comment]
`x_j` corresponds to the features which gets aggregated, and `x_i` denotes the features which aggregate. The reason `GATConv` uses both `x_i` and `x_j` is because the attention scores gets computed based on both source and target node features.

**Comment by MilkshakeForReal** [comment]
> 
> 
> `x_j` corresponds to the features of node `j` which gets aggregated, and `x_i` denotes the features of node `i` which aggregate. The reason `GATConv` uses both `x_i` and `x_j` is because the attention scores gets computed based on both nodes `i` and `j`.

Thank you very much for your timely reply!

When I studied the source code of GCNConv, I found that message function also needs x_i, because it is a sum of neighbor nodes and itself. But the reason, I think, why there's not x_i in the argument to the message function is because you've added  self-loops, which makes each node's neighbor contain itself, in other words, x_j contain x_i. Is my understanding correct?

Also, I saw in the source code for GTNConv that the self-loops were first removed and then added, I don't quite understand why to do so.

**Comment by rusty1s** [comment]
Yes and no. `x_i` and `x_j` are in edge parallel space, denoting the source and target node features for each edge. For aggregating the root node itself, we therefore have to add an additional self-loop. In `message`, you simply denote which message vector gets send to node `i` for each neighboring node `j`. This can be just based on the information given by `j` itself (GCNConv) or also based on node `i` (GATConv).

We remove self-loops first in case the user already provides self-loops in `edge_index`. Normally, the removal is a no-op.

**Comment by MilkshakeForReal** [comment]
> 
> 
> Yes and no. `x_i` and `x_j` are in edge parallel space, denoting the source and target node features for each edge. For aggregating the root node itself, we therefore have to add an additional self-loop. In `message`, you simply denote which message vector gets send to node `i` for each neighboring node `j`. This can be just based on the information given by `j` itself (GCNConv) or also based on node `i` (GATConv).
> 
> We remove self-loops first in case the user already provides self-loops in `edge_index`. Normally, the removal is a no-op.

Thank you very much. Now I realize that the reason I cannot understand it is because I cannot understand how the propagate function is implemented. I am not a very professional programmer. This leads me to be unclear about what I need to write myself when I try to define my own layer. I tried DGL and found that I could easily implement GCN and GAT. Maybe this library will be more suitable for me

**Comment by rusty1s** [comment]
The `MessagePassing` interface is really similar in both libraries. In PyG, `propagate` first computes messages for each edge `(j,i)` in the `message` function and aggregates those afterwards using on the of the aggregation operators `add`, `mean` or `max` for each neighborhood. In the `message` function, you have access to source and target node features and edge features to define messages, *.e.g.*,
```python
def message(self, x_j, x_i, edge_weight):
    return self.mlp(torch.cat([x_j, x_i], dim=1) * edge_weight.view(-1, 1)
```
which concatenates source and target node features, transforms them using a MLP and weights them based on edge weights.

**Comment by MilkshakeForReal** [comment]
Thank you for your patient answer! You must be an angel! I think I'm close to mastering the message function. However I still don't know the argument 'edge_index', 'edge_index_i', 'edge_index_j' and how do I know the arguments of self.propagate()?

**Comment by rusty1s** [comment]
edge_index describes the connectivity of your graph in a sparse COO format. This is needed as an argument to propagate in order to derive which messages should get passed to which nodes. This is the only required argument to propagate, all other ones are optional. You simply pass in all arguments there which are needed to create messages. edge_index_i is just a way to say derive all the target node indices from all the edges, and can be used to, e.g., normalize features around each neighborhood.

**Comment by MilkshakeForReal** [comment]
I think I will master it after studying more source code, thank you for your help!
BTW, I am currently dealing with a problem that requires updating the node features and updating the edge features. I noticed that the MessagePasing framework is based on updating node features. Is there any way I can add the update edge feature process to the model?
I think this process will become more and more important in the future for some tasks such as link prediction and link classification, etc.

**Comment by rusty1s** [comment]
Well, updating edge features is relatively easy using plain PyTorch methods, so an own interface is not really necessary. Simply do something like:
```python
row, col = edge_index
new_edge_attr = self.mlp(torch.cat([x[row], x[col], edge_attr], dim=1))
```

**Comment by MilkshakeForReal** [comment]
You are right, it is pretty easy, I just need to add this code to the forward function. I should have thought of it...

**Comment by NIRVANALAN** [comment]
In pytorch_geometric lib you always treat the `col` as target and `row` as source right? By default the message flows from `row` to `col` which the edge_index list can be denoted as `(j,i)`?
